File: jvm/core/main/scala/com/webank/eggroll/core/deepspeed/meta/store/RendezvousStoreSetRequest.scala
package com.webank.eggroll.core.deepspeed.meta.store

import com.google.protobuf.ByteString
import com.webank.eggroll.core.meta.Deepspeed

import scala.language.implicitConversions

case class RendezvousStoreSetRequest(prefix: String, key: K, value: V)

object RendezvousStoreSetRequest {
  implicit def serialize(src: RendezvousStoreSetRequest): Array[Byte] = {
    val builder = Deepspeed.StoreSetRequest.newBuilder()
      .setPrefix(src.prefix)
      .setKey(ByteString.copyFrom(src.key.toArray))
      .setValue(ByteString.copyFrom(src.value.toArray))
    builder.build().toByteArray
  }

  implicit def deserialize(byteString: ByteString): RendezvousStoreSetRequest = {
    val proto = Deepspeed.StoreSetRequest.parseFrom(byteString)
    RendezvousStoreSetRequest(
      prefix = proto.getPrefix,
      key = proto.getKey.toByteArray.toVector,
      value = proto.getValue.toByteArray.toVector
    )
  }
}

File: jvm/new_core/src/main/java/com/eggroll/core/deepspeed/store/RendezvousStoreSetRequest.java
package com.eggroll.core.deepspeed.store;

import com.google.protobuf.ByteString;
import com.google.protobuf.InvalidProtocolBufferException;
import com.webank.eggroll.core.meta.Deepspeed;
import lombok.Data;

@Data
public class RendezvousStoreSetRequest {
    private String prefix;
    private byte[] key;
    private byte[] value;

    public byte[] serialize() {
        Deepspeed.StoreSetRequest.Builder builder = Deepspeed.StoreSetRequest.newBuilder()
                .setPrefix(this.getPrefix())
                .setKey(ByteString.copyFrom(this.getKey()));
        builder.setValue(ByteString.copyFrom(this.getValue()));
        return builder.build().toByteArray();
    }

    public void deserialize(byte[] data) throws InvalidProtocolBufferException {
        Deepspeed.StoreSetRequest proto = Deepspeed.StoreSetRequest.parseFrom(data);
        this.prefix = proto.getPrefix();
        this.key = proto.getKey().toByteArray();
        this.value = proto.getValue().toByteArray();
    }
}